clarity/contracts/arkadiko-auction-engine-v4-1.clar

```diff
@@ -0,0 +1,396 @@
+    (asserts! (is-eq (contract-of oracle) (unwrap-panic (contract-call? .arkadiko-dao get-qualified-name-by-name "oracle"))) (err ERR-NOT-AUTHORIZED))
+    (asserts!
+      (or
+        (is-eq (contract-of reserve) (unwrap-panic (contract-call? .arkadiko-dao get-qualified-name-by-name "stx-reserve")))
+        (is-eq (contract-of reserve) (unwrap-panic (contract-call? .arkadiko-dao get-qualified-name-by-name "sip10-reserve")))
+      )
+      (err ERR-NOT-AUTHORIZED)
+    )
+    (asserts! 
+      (or
+        (and (is-eq token-string "STX") (is-eq (contract-of ft) .xstx-token))
+        (and (not (is-eq token-string "STX")) (is-eq (contract-of ft) coll-token-address))
+      ) 
+      (err ERR-TOKEN-TYPE-MISMATCH)
+    )
+    (asserts! (not (get is-liquidated vault)) (err ERR-AUCTION-NOT-ALLOWED))
+
+    (let (
+      (collateral-type (unwrap-panic (contract-call? vault-manager get-collateral-type-for-vault vault-id)))
+      (collateral-to-debt-ratio (unwrap-panic (contract-call? vault-manager calculate-current-collateral-to-debt-ratio vault-id coll-type oracle true)))
+      (liquidation-ratio (unwrap-panic (contract-call? coll-type get-liquidation-ratio collateral-type)))
+      (amounts (unwrap-panic (as-contract (contract-call? vault-manager liquidate vault-id coll-type))))
+
+      (updated-vault (contract-call? .arkadiko-vault-data-v1-1 get-vault-by-id vault-id))
+
+      (auction-id (+ (var-get last-auction-id) u1))
+      (auction {
+        id: auction-id,
+        auction-type: "collateral",
+        vault-id: vault-id,
+        collateral-amount: (get ustx-amount amounts),
+        collateral-token: (get collateral-token updated-vault),
+        collateral-address: (contract-of ft),
+        debt-to-raise: (get vault-debt amounts),
+        discount: (unwrap-panic (contract-call? coll-type get-liquidation-penalty (get collateral-type updated-vault))),
+        total-debt-burned: u0,
+        total-collateral-sold: u0,
+        ended-at: u0,
+      })
+    )
+      (asserts! (>= liquidation-ratio collateral-to-debt-ratio) (err ERR-NO-LIQUIDATION-REQUIRED))
+
+      ;; Add auction
+      (map-set auctions { id: auction-id } auction)
+
+      ;; Try to burn
+      (try! (burn-usda auction-id oracle coll-type vault-manager ft reserve liquidation-pool liquidation-rewards))
+
+      (print { type: "auction", action: "created", data: auction })
+      (ok true)
+    )
+  )
+)
+
+;; @desc burn USDA for vault and get part of collateral
+;; @param auction-id; the auction
+;; @param oracle; oracle
+;; @param coll-type; collateral types
+;; @param vault-manager; vault manager
+;; @param ft; collateral token if not STX vault
+;; @param reserve; reserve to get collateral tokens from
+;; @param liquidation-pool; pool to get USDA from
+;; @param liquidation-rewards; contract to deposit rewards
+(define-public (burn-usda
```

Comment:
REQ: this can be called on any auction, we should check that the auction is still open.

Reply:
If the auction is already closed, there is no collateral to sell and `usda-to-use` will be 0.